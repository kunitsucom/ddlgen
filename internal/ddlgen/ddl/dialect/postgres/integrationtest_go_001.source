package main

type (
	// User is a user.
	//
	// pgddl:      table: "users"
	// pgddl: constraint: UNIQUE("name")
	// pgddl:      index: CREATE INDEX "index_users_by_name" ON "users" ("name")
	User struct {
		// UserID is a user ID.
		UserID string `dbtest:"user_id" pgddl:"TEXT    NOT NULL" pkey:"true"`
		// Name is a user name.
		Name string   `dbtest:"name"    pgddl:"TEXT    NOT NULL"`
		// Email is a user email.
		Email string  `dbtest:"email"   pgddl:"TEXT    NOT NULL"`
		// Age is a user age.
		Age int       `dbtest:"age"     pgddl:"INTEGER NOT NULL"`
		// Ignore is a ignore field.
		Ignore string `dbtest:"-"       pgddl:"-"`
	}

	// Users is a user array.
	// This type is expected not to be detected.
	//
	// pgddl: table: "user_arrays"
	Users []*User

	// UserGroup is a user group.
	//
	// pgddl:table:
	UserGroup struct {
		// ID is a group ID.
		ID string   `dbtest:"id"   pgddl:"TEXT NOT NULL"`
		// Name is a group name.
		Name string `dbtest:"name" pgddl:"TEXT NOT NULL"`
	}

	// Author is a author.
	// pgddl:
	Author struct {
		// ID is a author ID.
		AuthorID string `dbtest:"id"     pgddl:"TEXT NOT NULL"`
		// Name is a author name.
		Name string     `dbtest:"name"   pgddl:"TEXT NOT NULL"`
	}

	// Book is a book.
	//
	// pgddl:table:"books"
	Book struct {
		// AuthorID is a book author.
		AuthorID string `pgddl:"TEXT NOT NULL" pkey:"true"`
		// ID is a book ID.
		ID string       `pgddl:"TEXT NOT NULL" pkey:"true"`
		// Title is a book title.
		Title string    `pgddl:"TEXT NOT NULL"`
	}

	// Store
	// pgddl: table: CREATE TABLE "stores"
	Store struct {
		// ID is a store ID.
		ID string   `dbtest:"id" pkey:"false"`
		// Name is a store name.
		Name string `dbtest:"name"`
	}
)

// pgddl: index: "index_books_by_title" ON "books" ("Title")
