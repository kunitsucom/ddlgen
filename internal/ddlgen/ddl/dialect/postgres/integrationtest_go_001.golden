-- Code generated by ddlgen. DO NOT EDIT.
--

-- source: postgres/integrationtest_go_001.source:6
-- User is a user.
--
-- pgddl:      table: "users"
-- pgddl: constraint: UNIQUE("name")
-- NOTE: the "User" struct's "Ignore" field has a tag for column name (`dbtest:"-"`), so the field is ignored.
CREATE TABLE "users" (
    -- UserID is a user ID.
    "user_id" TEXT    NOT NULL,
    -- Name is a user name.
    "name"    TEXT    NOT NULL,
    -- Email is a user email.
    "email"   TEXT    NOT NULL,
    -- Age is a user age.
    "age"     INTEGER NOT NULL,
    UNIQUE("name"),
    PRIMARY KEY ("user_id")
);

-- source: postgres/integrationtest_go_001.source:8
-- pgddl:      index: CREATE INDEX "index_users_by_name" ON "users" ("name")
CREATE INDEX "index_users_by_name" ON "users" ("name");

-- source: postgres/integrationtest_go_001.source:25
-- Users is a user array.
-- This type is expected not to be detected.
--
-- pgddl: table: "user_arrays"
-- ERROR: the comment (postgres/integrationtest_go_001.source:22) does not have struct fields for column type and constraint (`pgddl:"<TYPE> [CONSTRAINT]"`), or the comment is not associated with struct.
CREATE TABLE "user_arrays" (
);

-- source: postgres/integrationtest_go_001.source:30
-- Group is a group.
--
-- pgddl:table: Groups
CREATE TABLE Groups (
    -- ID is a group ID.
    "id"   TEXT NOT NULL,
    -- Name is a group name.
    "name" TEXT NOT NULL
);

-- source: postgres/integrationtest_go_001.source:39
-- Author is a author.
-- pgddl:
-- WARN: the comment (postgres/integrationtest_go_001.source:38) does not have a key for table (pgddl: table: CREATE TABLE <table>), so the struct name "Author" is used as the table name.
CREATE TABLE Author (
    -- ID is a author ID.
    "id"   TEXT NOT NULL,
    -- Name is a author name.
    "name" TEXT NOT NULL
);

-- source: postgres/integrationtest_go_001.source:49
-- Book is a book.
--
-- pgddl:table:"books"
CREATE TABLE "books" (
    -- WARN: the "Book" struct's "AuthorID" field does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "AuthorID" is used as the column name.
    -- AuthorID is a book author.
    "AuthorID" TEXT NOT NULL,
    -- WARN: the "Book" struct's "ID" field does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "ID" is used as the column name.
    -- ID is a book ID.
    "ID"       TEXT NOT NULL,
    -- WARN: the "Book" struct's "Title" field does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "Title" is used as the column name.
    -- Title is a book title.
    "Title"    TEXT NOT NULL,
    PRIMARY KEY ("AuthorID", "ID")
);

-- source: postgres/integrationtest_go_001.source:60
-- Store
-- pgddl: table: CREATE TABLE "stores"
CREATE TABLE "stores" (
    -- ERROR: the "Store" struct's "ID" field does not have a tag for column type and constraint (`pgddl:"<TYPE> [CONSTRAINT]"`)
    -- WARN: the field "ID" does not have valid primary key tag (`pkey:"true"`), so the column is not used as primary key.
    -- ID is a store ID.
    "id"   DDLGEN_ERROR_STRUCT_FIELD_TAG_NOT_FOUND,
    -- ERROR: the "Store" struct's "Name" field does not have a tag for column type and constraint (`pgddl:"<TYPE> [CONSTRAINT]"`)
    -- Name is a store name.
    "name" DDLGEN_ERROR_STRUCT_FIELD_TAG_NOT_FOUND
);

-- source: postgres/integrationtest_go_001.source:69
-- pgddl: index: "index_books_by_title" ON "books" ("Title")
CREATE INDEX "index_books_by_title" ON "books" ("Title");

-- source: postgres/integrationtest_go_001.source:69
-- WARN: the comment (postgres/integrationtest_go_001.source:69) does not have a key for table (pgddl: table: CREATE TABLE <table>), or the comment is not associated with struct.
